File: backend/src/config/storage.rs
use std::env;
use std::path::PathBuf;

fn env_bool(name: &str, default: bool) -> bool {
    env::var(name)
        .ok()
        .and_then(|value| {
            let trimmed = value.trim();
            if trimmed.is_empty() {
                None
            } else {
                trimmed.parse::<bool>().ok()
            }
        })
        .unwrap_or(default)
}

fn env_string(name: &str) -> Option<String> {
    env::var(name)
        .ok()
        .map(|value| value.trim().to_string())
        .filter(|v| !v.is_empty())
}

#[derive(Debug, Clone)]
pub struct MediaStorageConfig {
    pub upload_root: PathBuf,
    pub upload_base_url: Option<String>,
    pub public_base_url: Option<String>,
    pub bucket: Option<String>,
    pub region: Option<String>,
    pub endpoint: Option<String>,
    pub access_key: Option<String>,
    pub secret_key: Option<String>,
    pub session_token: Option<String>,
    pub force_path_style: bool,
    pub keep_local_copy: bool,
    pub remove_source_after_upload: bool,
}

impl MediaStorageConfig {
    pub fn from_env() -> Self {
        let upload_root = env::var("UPLOAD_STORAGE_PATH")
            .map(PathBuf::from)
            .unwrap_or_else(|_| PathBuf::from("uploads"));

        let upload_base_url = env_string("UPLOAD_BASE_URL");
        let public_base_url = env_string("PUBLIC_BASE_URL");

        let bucket = env_string("S3_BUCKET").or_else(|| env_string("AWS_S3_BUCKET"));
        let region = env_string("S3_REGION").or_else(|| env_string("AWS_REGION"));
        let endpoint = env_string("S3_ENDPOINT").or_else(|| env_string("AWS_S3_ENDPOINT"));

        let access_key = env_string("S3_ACCESS_KEY").or_else(|| env_string("AWS_ACCESS_KEY_ID"));
        let secret_key =
            env_string("S3_SECRET_KEY").or_else(|| env_string("AWS_SECRET_ACCESS_KEY"));
        let session_token =
            env_string("S3_SESSION_TOKEN").or_else(|| env_string("AWS_SESSION_TOKEN"));

        let force_path_style = env_bool("S3_FORCE_PATH_STYLE", endpoint.is_some());
        let keep_local_copy = env_bool("S3_KEEP_LOCAL_COPY", true);
        let remove_source_after_upload = env_bool("S3_REMOVE_SOURCE_AFTER_UPLOAD", true);

        Self {
            upload_root,
            upload_base_url,
            public_base_url,
            bucket,
            region,
            endpoint,
            access_key,
            secret_key,
            session_token,
            force_path_style,
            keep_local_copy,
            remove_source_after_upload,
        }
    }

    pub fn has_remote_backend(&self) -> bool {
        self.bucket.is_some() && self.access_key.is_some() && self.secret_key.is_some()
    }
}

File: video-renderer/src/config/storage.ts
import type { S3ClientConfig } from '@aws-sdk/client-s3';

const truthy = new Set(['1', 'true', 'yes', 'on']);

const env = (name: string): string | undefined => {
    const value = process.env[name];
    if (!value) {
        return undefined;
    }
    const trimmed = value.trim();
    return trimmed.length > 0 ? trimmed : undefined;
};

const envBool = (name: string, fallback = false): boolean => {
    const value = env(name);
    if (!value) {
        return fallback;
    }
    const normalized = value.toLowerCase();
    if (truthy.has(normalized)) {
        return true;
    }
    if (['0', 'false', 'no', 'off'].includes(normalized)) {
        return false;
    }
    return fallback;
};

export type RendererStorageConfig = {
    bucket: string;
    region?: string;
    endpoint?: string;
    accessKeyId?: string;
    secretAccessKey?: string;
    sessionToken?: string;
    forcePathStyle: boolean;
    storagePrefix: string;
    uploadPrefix: string;
    publicBaseUrl?: string;
};

export const isS3UploadEnabled = (): boolean => {
    return (
        envBool('RENDERER_S3_UPLOAD', false) ||
        envBool('VIDEO_RENDERER_S3_UPLOAD', false) ||
        envBool('VIDEO_RENDERER_UPLOAD_TO_S3', false)
    );
};

export const loadStorageConfig = (): RendererStorageConfig | null => {
    const bucket =
        env('S3_BUCKET') ??
        env('AWS_S3_BUCKET');

    if (!bucket) {
        return null;
    }

    const region = env('S3_REGION') ?? env('AWS_REGION');
    const endpoint = env('S3_ENDPOINT') ?? env('AWS_S3_ENDPOINT');
    const accessKeyId = env('S3_ACCESS_KEY') ?? env('AWS_ACCESS_KEY_ID');
    const secretAccessKey =
        env('S3_SECRET_KEY') ?? env('AWS_SECRET_ACCESS_KEY');
    const sessionToken = env('S3_SESSION_TOKEN') ?? env('AWS_SESSION_TOKEN');
    const forcePathStyle = envBool(
        'S3_FORCE_PATH_STYLE',
        Boolean(endpoint)
    );

    const storagePrefix =
        env('S3_STORAGE_PREFIX') ??
        env('MEDIA_STORAGE_PREFIX') ??
        'uploads';
    const uploadPrefix =
        env('S3_RENDER_UPLOAD_PREFIX') ??
        env('VIDEO_RENDERER_UPLOAD_PREFIX') ??
        'renders';
    const publicBaseUrl =
        env('UPLOAD_BASE_URL') ??
        env('PUBLIC_BASE_URL');

    return {
        bucket,
        region,
        endpoint,
        accessKeyId,
        secretAccessKey,
        sessionToken,
        forcePathStyle,
        storagePrefix: storagePrefix.replace(/^\/+|\/+$/g, ''),
        uploadPrefix: uploadPrefix.replace(/^\/+|\/+$/g, ''),
        publicBaseUrl: publicBaseUrl?.trim()
    };
};

export const buildS3ClientConfig = (
    config: RendererStorageConfig
): S3ClientConfig => {
    const clientConfig: S3ClientConfig = {};

    if (config.region) {
        clientConfig.region = config.region;
    }

    if (config.accessKeyId && config.secretAccessKey) {
        clientConfig.credentials = {
            accessKeyId: config.accessKeyId,
            secretAccessKey: config.secretAccessKey,
            sessionToken: config.sessionToken
        };
    }

    if (config.endpoint) {
        clientConfig.endpoint = config.endpoint;
    }

    if (config.forcePathStyle) {
        clientConfig.forcePathStyle = true;
    }

    return clientConfig;
};


